homeassistant/components/vesync/diagnostics.py

```diff
@@ -0,0 +1,160 @@
+"""Diagnostics support for VeSync."""
+from __future__ import annotations
+
+from typing import Any
+
+from pyvesync import VeSync
+
+from homeassistant.components.diagnostics import REDACTED
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.device_registry import DeviceEntry
+
+from .common import VeSyncBaseDevice
+from .const import DOMAIN, VS_MANAGER
+
+KEYS_TO_REDACT = {"manager"}
+
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+    manager: VeSync = hass.data[DOMAIN][VS_MANAGER]
+
+    data = {}
+    data.update(
+        {
+            DOMAIN: {
+                "account_id": manager.account_id,
+                "bulb_count": len(manager.bulbs),
+                "fan_count": len(manager.fans),
+                "outlets_count": len(manager.outlets),
+                "switch_count": len(manager.switches),
+                "timezone": manager.time_zone,
+            }
+        }
+    )
+
+    data.update(
+        {
+            "home_assistant": {
+                "disabled_by": entry.disabled_by,
+                "disabled_polling": entry.pref_disable_polling,
+            }
+        }
+    )
+
+    data.update(
+        {
+            "devices": {
+                "bulbs": [
+                    _async_device_as_dict(hass, None, device)
+                    for device in manager.bulbs
+                ],
+                "fans": [
+                    _async_device_as_dict(hass, None, device) for device in manager.fans
+                ],
+                "outlets": [
+                    _async_device_as_dict(hass, None, device)
+                    for device in manager.outlets
+                ],
+                "switches": [
+                    _async_device_as_dict(hass, None, device)
+                    for device in manager.switches
+                ],
+            }
+        }
+    )
+
+    return data
+
+
+async def async_get_device_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry, device: DeviceEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a device entry."""
+    manager: VeSync = hass.data[DOMAIN][VS_MANAGER]
+
+    device_dict = _build_device_dict(manager)
+    vesync_device_id = next(iter(device.identifiers))[1]
+    return _async_device_as_dict(hass, device, device_dict[vesync_device_id])
+
+
+@callback
+def _async_device_as_dict(
+    hass: HomeAssistant,
+    hass_device: DeviceEntry | None,
+    vesync_device: VeSyncBaseDevice,
+) -> dict[str, Any]:
+    """Represent a VeSync device as a dictionary."""
+
+    # Base device information, without sensitive information.
+    data = _redact_device_values(vesync_device)
+
+    if hass_device:
```

Comment:
this makes no sense. How could it happen to be None? And if that's possible, don't pass it to this function.

Reply:
You are correct, an early version of this file used this method for both diagnostics paths (config and device).  Later refactorings meant that the only value for the other path was the redaction.  This flow has been changed.